### DISCUSSION TITLE
MsSqlBuilder - why is WithDatabase private?

### DISCUSSION THREAD

**Discussion by pborgmans** [testcontainers/testcontainers-dotnet#828] [Q&A]
Is there a reason why WithDatabase is declared private on the MsSqlBuilder? 
I would like to work in a self named database instead of master to execute my EF Core migrations.

Thanks in advance!

**Comment by HofmeisterAn** [comment]
> Is there a reason why WithDatabase is declared private on the MsSqlBuilder?

Yes, the [Microsoft SQL Server](https://hub.docker.com/_/microsoft-mssql-server) image does not support configuring a database name by default. To create a custom database, you can follow the steps mentioned [here](https://github.com/microsoft/mssql-docker/tree/master/linux/preview/examples/mssql-customize). You need to copy the scripts via [WithResourceMapping](https://github.com/testcontainers/testcontainers-dotnet/blob/17f04940412468344235065038929cf731f871ae/src/Testcontainers/Builders/ContainerBuilder%603.cs#L165-L181) to the container before it starts. Or use as mentioned EF or a client library.

**Comment by fretje** [comment]
I'm also interested to know why this is not supported.

Also, the link you're referring to, to "create a custom database" talks about building a whole new image. I want to use the standard sqlserver 2022 image (which is perfectly possible with the "WithImage" method), simply with another database name (which was perfectly possible with TestContainers 2.3.0 with the "WithDatabase" method).

Having to build a whole new docker image, just for that, seems like a bit overkill. 
Also seems like a regression from the previous version?

**Reply by HofmeisterAn** [reply]
> Having to build a whole new docker image, just for that, seems like a bit overkill.

You do not need to build a new Docker image. As mentioned, you can provide the scripts before the container starts and override the entrypoint. There is no need to build the image from scratch.

> TestContainers 2.3.0 with the "WithDatabase" method

This was possible because we run a custom SQL [script](https://github.com/testcontainers/testcontainers-dotnet/blob/83967725b9d03f7d660e829a6575c44e8759f7ab/src/Testcontainers/Containers/Modules/Databases/MsSqlTestcontainer.cs#L59-L67). Something you still can do by using [ExecScriptAsync(string)](https://github.com/testcontainers/testcontainers-dotnet/blob/17f04940412468344235065038929cf731f871ae/tests/Testcontainers.MsSql.Tests/MsSqlContainerTest.cs#L39-L40).

We have begun aligning more with other language implementations, and as such, we are taking into account the experiences of Java. Up until now, we have seen that configuring the database has been the responsibility of the client library or a database migration tool. It probably does not make sense to maintain client library functionalities in Testcontainers.

Personally, I am torn. I can understand that the feature request is somewhat of a no-brainer, but on the other hand, any more mature implementation will likely use the client library or a database migration tool regardless. It would be great if the functionality, is provided by the image itself instead of having the feature in Testcontainers (e.g. like PostgreSQL).

Therefore, I have started a discussion on the following link: https://github.com/testcontainers/testcontainers-dotnet/discussions/801. Please consider sharing your feedback and thoughts.

**Reply by moccajoghurt** [reply]
Does this mean that I have to embed these 3 files into my test project?
configure-db.sh
entrypoint.sh
setup.sql
And then use the functions "WithResourceMapping" and "WithEntrypoint" just to create a database?
Nobody will understand what's going on when they see my integration test. I'll have to point them to this discussion :(

**Reply by HofmeisterAn** [reply]
> Does this mean that I have to embed these 3 files into my test project?

Yes, or an implementation of `IResourceMapping`. There are various patterns and mechanisms available to create a custom database, this is not the only option. Typically, a database migration tool is used. E.g. if you are using EF, all you need to do is:

```csharp
var connectionString = new SqlConnectionStringBuilder(weatherForecastTest._msSqlContainer.GetConnectionString());
connectionString.InitialCatalog = Guid.NewGuid().ToString("D"); // New database name, EF will create it.
```

Make sure you are calling `Database.EnsureCreated()` in your `DbContext` implementation.

> And then use the functions "WithResourceMapping" and "WithEntrypoint" just to create a database?

Exactly, but this allows executing any SQL script (not just creating the database). As I suggested in https://github.com/testcontainers/testcontainers-dotnet/pull/855#issuecomment-1491389100, we can integrate this feature into Testcontainers. This enables developers to map their own custom SQL scripts and execute them before the container indicates readiness too, similar to the PostgreSQL [image](https://hub.docker.com/_/postgres).

> If you would like to do additional initialization in an image derived from this one, add one or more *.sql, *.sql.gz, or *.sh scripts under /docker-entrypoint-initdb.d (creating the directory if necessary). After the entrypoint calls initdb to create the default postgres user and database, it will run any *.sql files, run any executable *.sh scripts, and source any non-executable *.sh scripts found in that directory to do further initialization before starting the service.

On top of that we can expose a `WithDatabase(string)` API. I would like to reiterate that this limitation is inherent to the MSSQL image, and not a problem specific to Testcontainers.

**Reply by chrisrlewis** [reply]
I have stumbled across this thread while trying to get an MsSqlContainer up and running for integration testing. I have spent some time wondering why my db tables and migrations have been appearing in the 'master' database and why it doesn't seem possible to provide a database name. 

I've read the above, and considered the reasoning and looked at the https://github.com/microsoft/mssql-docker/tree/80e2a51d0eb1693f2de014fb26d4a414f5a5add5/linux/preview/examples/mssql-customize thread and #801.

For what it's worth, from my perspective I just want to spin up a new database with a supplied name, and I want to be able to do this simply and quickly. In my view, usability is the most important consideration when using third party libraries and tools. The typical project may use very many libraries, etc. and we time-poor software engineers don't have time to delve into finer intricacies of achieving basic things. The code also needs to be readable and understandable for anyone who needs to read it. 

Can we have a .WithDatabaseName("mydatabase") for doing this? I've seen this has been available on a JDBC module, and the same for MsSql would be very welcome.

**Reply by HofmeisterAn** [reply]
I can only reference previous discussions and emphasize that I am in favor of supporting this feature. However, calling `WithDatabase(string)` must set up container instances with the desired database rather than simply modifying the connection string to a non-existent database. The instances should be immediately usable without relying on additional user configurations, setups, or frameworks. Therefore, I propose enhancing the MSSQL module by overriding the default entrypoint to enable the execution of custom SQL scripts. This can be utilized to create the desired custom database and make `WithDatabase(string)` public.

- https://github.com/testcontainers/testcontainers-dotnet/issues/986#issuecomment-1700462953
- https://github.com/testcontainers/testcontainers-dotnet/issues/986#issuecomment-1700524225
- https://github.com/testcontainers/testcontainers-dotnet/issues/1029#issuecomment-1769083569
- https://github.com/testcontainers/testcontainers-dotnet/issues/1029#issuecomment-1769313037

**Reply by chrisrlewis** [reply]
Thanks for your swift reply @HofmeisterAn. I appreciate that there is complexity and a lot to consider outside of my simple use case. 

What you mention above (modifying the connection string for a non-existent database) has worked for my use case, though I hear what you say about this being potentially problematic. In my case, I'm using EF Core so - at least at first look - it does give me what I need. 

I will keep a look out for updates on the repo. Do you have a timescale for the changes?

**Reply by HofmeisterAn** [reply]
> In my case, I'm using EF Core so - at least at first look - it does give me what I need.

Exactly. It works for someone using EF, but not for someone relying on a plain SQL connection. The changes I propose are not rocket science. If someone really needs it, it can be implemented quite quickly (and contributed). However, as I mentioned, I do not plan to do it using my free resources soon. It is not blocking anyone from using the module at all.

> Do you have a timescale for the changes?

No, not really. Maybe if I am in the mood during the upcoming holidays. Otherwise, sometime in 2024 😅.

**Comment by iozcelik** [comment]
Yes, @fretje is right. It is possible to rename. I think it will be change to public and support database naming.
